sqlparse/keywords.py

```diff
@@ -77,15 +105,56 @@ def is_keyword(value):
         (r'CREATE(\s+OR\s+REPLACE)?\b', tokens.Keyword.DDL),
         (r'DOUBLE\s+PRECISION\b', tokens.Name.Builtin),
 
-        (r'[0-9_A-ZÀ-Ü][_$#\w]*', is_keyword),
+        (r'[0-9_A-ZÀ-Ü][_$#\w]*', is_general_SQL_keyword),
 
         (r'[;:()\[\],\.]', tokens.Punctuation),
         (r'[<>=~!]+', tokens.Operator.Comparison),
         (r'[+/@#%^&|`?^-]+', tokens.Operator),
-    ]}
+    ],
+    'TransactSQL': [
+        (r'(\r\n|\r|\n)', tokens.Newline),
+        (r'\s+', tokens.Whitespace),
+
+        (r'--(?m).*?$\n?', tokens.Comment.Single),
+        (r'/\*[\s\S]*?\*/', tokens.Comment.Multiline),
+
+        (r'0x[0-9a-f]+', tokens.Number.Hexadecimal),
+        (r'-?\d*(\.\d+)?E-?\d+', tokens.Number.Float),
+        (r'-?(\d+(\.\d*)|\.\d+)', tokens.Number.Float),
+        (r'-?[0-9]+', tokens.Number.Integer),
+
+        (r"'(''|\\\\|\\'|[^'])*'", tokens.String.Single),
+        (r"'(''|[^'])*'", tokens.String.Single),
+        (r'"(""|[^"])*"', tokens.String.Symbol),
+
+        (r'\*', tokens.Wildcard),
+        (r'[A-ZÀ-Ü]\w*(?=\s*\.)', tokens.Name),  # 'Name'   .
+        (r'(?<=\.)[A-ZÀ-Ü]\w*', tokens.Name),  # .'Name'
+        (r'((LEFT\s+|RIGHT\s+|FULL\s+)?(INNER\s+|OUTER\s+)?'
+         r'|(CROSS\s+)?)?JOIN\b', tokens.Keyword),
+        (r'NOT\s+NULL\b', tokens.Keyword),
+        (r'UNION\s+ALL\b', tokens.Keyword),
+        (r'[A-ZÀ-Ü]\w*(?=\()', tokens.Name),  # side effect: change kw to func
+        # Below we use \w even for the first "real" character because
+        # tokens starting with a digit have already been recognized
+        # as Number above.
+        (r'@@\w+', tokens.Name.Builtin),
+
+        (r'[0-9_A-ZÀ-Ü][_$#\w]*', is_TransactSQL_keyword),
+        (r'(?<![\w\])])(\[[^\]]+\])', tokens.Name),
+        (r'@\w+', tokens.Name.Variable),
+        (r'[;(),.]', tokens.Punctuation),
+        (r'#?#?\w+', tokens.Name),  # names for temp tables and anything else
+        (r'\?', tokens.Name.Variable.Magic),
+        (r'[<>=~!]+', tokens.Operator.Comparison),
+        (r'[+/@#%^&|?^-]+', tokens.Operator),
+    ]
+}
 
 FLAGS = re.IGNORECASE | re.UNICODE
-SQL_REGEX = [(re.compile(rx, FLAGS).match, tt) for rx, tt in SQL_REGEX['root']]
+
+# Leave empty so the client can add additional keywords to parse against via the parser options.
+ADDITIONAL_KEYWORDS = {}
```

Comment:
I think we should avoid a dictionary on module level that is changed by a function (update_additional_keywords). When using the module programmatically to parse multiple statements ADDITIONAL_KEYWORDS will not be reset after each run. So a first call with additional_keywords in options and a subsequent call or a different thread (both in the same process) without will still use the additional keywords from the first run.

Maybe we should treat additional keywords differently and not using them in the _is_keyword(value) functions. I'm not sure what side-effects a change of the _is_keyword() functions would introduce.